### PR TITLE
Add support for non-persistent buffers.

### PR DESCRIPTION
**Pull request by sharvil**
Issue: #18056

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
@SsnL Do you have time to review this PR? Thanks

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -105,19 +106,23 @@ def forward(self, *input):
...
-    def register_buffer(self, name, tensor):
-        r"""Adds a persistent buffer to the module.
+    def register_buffer(self, name, tensor, persistent=True):
+        r"""Adds a buffer to the module.
 
         This is typically used to register a buffer that should not to be
         considered a model parameter. For example, BatchNorm's ``running_mean``
-        is not a parameter, but is part of the persistent state.
+        is not a parameter, but is part of the module's state. Buffers, by
+        default, are persistent and will be saved alongside parameters. This
```
Comment (albanD) [inline torch/nn/modules/module.py:115]: Could you make this clearer that the only difference here is that non persistent buffers won't show up in the state dict load/save while the persistent ones will.
There is no other difference right?

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -85,6 +85,7 @@ def __init__(self):
         self.training = True
         self._parameters = OrderedDict()
         self._buffers = OrderedDict()
+        self._persistent_buffers = set()
```
Comment (albanD) [inline torch/nn/modules/module.py:88]: nit: to make this clearer that this is not a dict like the other things, you could name this `self._persistent_buffers_name_set` ?  Or is this getting too long?

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -618,6 +627,7 @@ def remove_from(*dicts):
                 raise AttributeError(
                     "cannot assign parameters before Module.__init__() call")
             remove_from(self.__dict__, self._buffers, self._modules)
+            self._persistent_buffers.discard(name)
```
Comment (albanD) [inline torch/nn/modules/module.py:630]: nit: I think you can add the discard call inside the `remove_from` function here to make this slightly simpler.

**Review by albanD — CHANGES_REQUESTED** [review]
Hi,

Sorry for the late review.
Could you make a small update in the doc to make it crystal clear what this can be used for?
The code is good, only minor comments.

File: torch/nn/modules/module.py (hunk 2 above)
Comment (sharvil) [reply torch/nn/modules/module.py]: Done. I'm not a huge fan of appending the type name to the identifier, but don't feel strongly about it.

File: torch/nn/modules/module.py (hunk 1 above)
Comment (sharvil) [reply torch/nn/modules/module.py:115]: Done. Yes, there's no other difference.

File: torch/nn/modules/module.py (hunk 3 above)
Comment (sharvil) [reply torch/nn/modules/module.py]: Done.

**Review by sharvil — COMMENTED** [review]
Thanks for the code review, @albanD!

**Review by albanD — APPROVED** [review]
Perfect thanks!

**Comment by sharvil** [comment]
(last push was just a rebase to pytorch/master)

**Comment by albanD** [comment]
Ho nice, I was going to do it manually but better that way. Thanks!

**Pull request closed without merging**